File: BookApp/BookApp/Models/BookState.swift
//
//  BookState.swift
//  BookApp
//
//  Created by Andrii Tishchenko on 2024-08-27.
//

import Foundation

struct BookState {
    var books: [Book]
    var favorites: [Int]
    var searchResults: [Book]
    let isLoading: Bool
    let error: String?
    var routing: RoutingState
    
    static var initial: BookState {
        return BookState(books: [], favorites: [], searchResults: [], isLoading: false, error: nil, routing: .home)
    }
}

extension BookState {
    func copy(
        books: [Book]? = nil,
        favorites: [Int]? = nil,
        searchResults: [Book]? = nil,
        isLoading: Bool? = nil,
        error: String? = nil,
        routing: RoutingState? = nil
    ) -> BookState {
        return BookState(
            books: books ?? self.books,
            favorites: favorites ?? self.favorites,
            searchResults: searchResults ?? self.searchResults,
            isLoading: isLoading ?? self.isLoading,
            error: error ?? self.error,
            routing: routing ?? self.routing
        )
    }
}

File: BooksDemo/app/src/main/java/com/example/booksdemo/intent/BookState.kt
package com.example.booksdemo.intent

import com.example.booksdemo.models.Book

data class BookState(
    val books: List<Book> = emptyList(),
    val favorites: List<Int> = emptyList(),
    val searchResults: List<Book> = emptyList(),
    val isLoading: Boolean = true,
    val error: String? = "Error message",
    val routing: Screen = Screen.Home
)
